app/src/main/java/com/mapbox/navigation/examples/basics/MultipleWayPointsActivity.kt

```diff
@@ -0,0 +1,556 @@
+
+    /**
+     * Generates updates for the [MapboxManeuverView] to display the upcoming maneuver instructions
+     * and remaining distance to the maneuver point.
+     */
+    private lateinit var maneuverApi: MapboxManeuverApi
+
+    /**
+     * [NavigationLocationProvider] is a utility class that helps to provide location updates generated by the Navigation SDK
+     * to the Maps SDK in order to update the user location indicator on the map.
+     */
+    private val navigationLocationProvider = NavigationLocationProvider()
+
+    /**
+     * Gets notified whenever the tracked routes change.
+     *
+     * A change can mean:
+     * - routes get changed with [MapboxNavigation.setRoutes]
+     * - routes annotations get refreshed (for example, congestion annotation that indicate the live traffic along the route)
+     * - driver got off route and a reroute was executed
+     */
+    private val routesObserver = RoutesObserver { routeUpdateResult ->
+        if (routeUpdateResult.routes.isNotEmpty()) {
+            // generate route geometries asynchronously and render them
+            val routeLines = routeUpdateResult.routes.map { RouteLine(it, null) }
+
+            routeLineApi.setRoutes(
+                routeLines
+            ) { value ->
+                mapboxMap.getStyle()?.apply {
+                    routeLineView.renderRouteDrawData(this, value)
+                }
+            }
+
+            // update the camera position to account for the new route
+            viewportDataSource.onRouteChanged(routeUpdateResult.routes.first())
+            viewportDataSource.evaluate()
+        } else {
+            // remove the route line and route arrow from the map
+            val style = mapboxMap.getStyle()
+            if (style != null) {
+                routeLineApi.clearRouteLine { value ->
+                    routeLineView.renderClearRouteLineValue(
+                        style,
+                        value
+                    )
+                }
+            }
+
+            // remove the route reference from camera position evaluations
+            viewportDataSource.clearRouteData()
+            viewportDataSource.evaluate()
+        }
+    }
+
+    /**
+     * Gets notified with progress along the currently active route.
+     */
+    private val routeProgressObserver = RouteProgressObserver { routeProgress ->
+        // update the camera position to account for the progressed fragment of the route
+        viewportDataSource.onRouteProgressChanged(routeProgress)
+        viewportDataSource.evaluate()
+
+        // update top banner with maneuver instructions
+        val maneuvers = maneuverApi.getManeuvers(routeProgress)
+        maneuvers.fold(
+            { error ->
+                Toast.makeText(
+                    this@MultipleWaypointsActivity,
+                    error.errorMessage,
+                    Toast.LENGTH_SHORT
+                ).show()
```

Comment:
I'm constantly getting `index 1 size 0` error message. Can't find more data about in the logs. Any ideas how to troubleshoot?

Reply:
Oh, my bad. it's not a native code. I can debug it 🙂

Reply:
`ManeuverProcessor.updateDistanceRemainingForCurrentManeuver` throws an index out bounds exception because there is an empty array`stepToManeuvers.maneuverList` and the `currentInstructionIndex` is 1.

Reply:
please cut a ticket, it looks like data issue

Reply:
https://github.com/mapbox/mapbox-navigation-android/issues/5088